pkg/kubemark/cloudprovider.go

```diff
@@ -0,0 +1,368 @@
+// respective replication controllers have been deleted and the nodes have
+// become unready. This is to cover for the fact that there is no proper cloud
+// provider for kubemark that would care for deleting the nodes.
+type kubemarkCluster struct {
+	client            kubeclient.Interface
+	nodeLister        informersv1.NodeInformer
+	nodesToDelete     map[string]bool
+	nodesToDeleteLock sync.Mutex
+}
+
+// NewProvider creates Provider using the privided clients to talk to external
+// and kubemark clusters. It returns an error if it fails to sync data from
+// any of the two clusters.
+func NewProvider(externalClient kubeclient.Interface, kubemarkClient kubeclient.Interface, stop <-chan struct{}) (*Provider, error) {
+	externalInformerFactory := informers.NewSharedInformerFactory(externalClient, 0)
+	kubemarkInformerFactory := informers.NewSharedInformerFactory(kubemarkClient, 0)
+	manager := &Provider{
+		externalCluster: externalCluster{
+			rcLister:  externalInformerFactory.InformerFor(&apiv1.ReplicationController{}, newReplicationControllerInformer),
+			podLister: externalInformerFactory.InformerFor(&apiv1.Pod{}, newPodInformer),
+			client:    externalClient,
+		},
+		kubemarkCluster: kubemarkCluster{
+			nodeLister:        kubemarkInformerFactory.Core().V1().Nodes(),
+			client:            kubemarkClient,
+			nodesToDelete:     make(map[string]bool),
+			nodesToDeleteLock: sync.Mutex{},
+		},
+	}
+
+	externalInformerFactory.Start(stop)
+	for _, synced := range externalInformerFactory.WaitForCacheSync(stop) {
+		if !synced {
+			return nil, fmt.Errorf("failed to sync data of external cluster")
+		}
+	}
+	manager.kubemarkCluster.nodeLister.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: manager.kubemarkCluster.removeUnneededNodes,
+	})
+	go manager.kubemarkCluster.nodeLister.Informer().Run(stop)
+	if !manager.kubemarkCluster.nodeLister.Informer().HasSynced() {
+		return nil, fmt.Errorf("failed to sync data of kubemark cluster")
+	}
+
+	rand.Seed(time.Now().UTC().UnixNano())
+
+	// Get hollow node template from an existing hollow node to be able to create
+	// new nodes based on it.
+	nodeTemplate, err := manager.getNodeTemplate()
+	if err != nil {
+		return nil, err
+	}
+	manager.nodeTemplate = nodeTemplate
+	return manager, nil
+}
+
+// GetNodeGroupForNode returns the name of the node group to which node belongs.
+// Returns an error if node group for node was not found.
+func (kubemarkProvider *Provider) GetNodeGroupForNode(node string) (string, error) {
+	for _, podObj := range kubemarkProvider.externalCluster.podLister.GetStore().List() {
+		pod, ok := podObj.(*apiv1.Pod)
+		if !ok {
+			continue
+		}
+		if pod.ObjectMeta.Name == node {
+			nodeGroup, ok := pod.ObjectMeta.Labels[nodeGroupLabel]
+			if ok {
+				return nodeGroup, nil
+			}
+			return "", fmt.Errorf("can't find nodegroup for node %s. Node exists but does not have the %s label", nodeGroupLabel, node)
```

Comment:
Maybe make the error msg shorter? Something like "can't find nodegroup for node %s missing the label '%s' "